src/Native/Unix/System.Native/pal_console.c

```diff
@@ -71,92 +72,123 @@ void SystemNative_SetKeypadXmit(const char* terminfoString)
+        termios->c_lflag = g_initTermios.c_lflag;
+    }
+    else
+    {
+        termios->c_iflag &= (uint32_t)(~(IXON | IXOFF));
+        termios->c_lflag &= (uint32_t)(~(ECHO | ICANON | IEXTEN));
     }
 }
 
-static void IncorporateBreak(struct termios *termios, int32_t signalForBreak)
+static bool TcGetAttr(struct termios* termios, bool signalForBreak, bool hasInteractiveChildren)
 {
-    assert(termios != NULL);
-    assert(signalForBreak == 0 || signalForBreak == 1);
+    bool rv = tcgetattr(STDIN_FILENO, termios) >= 0;
+    if (rv)
+    {
+        ApplyTerminalSettings(termios, signalForBreak, hasInteractiveChildren);
+    }
+    return rv;
+}
 
-    if (signalForBreak)
-        termios->c_lflag |= (uint32_t)ISIG;
-    else
-        termios->c_lflag &= (uint32_t)(~ISIG);
+static bool TcSetAttr(struct termios* t)
+{
+    // This method is called with lock g_lock held.
+
+    if (g_consoleUninitialized)
+    {
+        // The application is exiting, we mustn't change terminal settings.
+        return true;
+    }
+
+    return tcsetattr(STDIN_FILENO, TCSANOW, t) >= 0;
 }
 
-// In order to support Console.ReadKey(intercept: true), we need to disable echo and canonical mode.
-// We have two main choices: do so for the entire app, or do so only while in the Console.ReadKey(true).
-// The former has a huge downside: the terminal is in a non-echo state, so anything else that runs
-// in the same terminal won't echo even if it expects to, e.g. using Process.Start to launch an interactive,
-// program, or P/Invoking to a native library that reads from stdin rather than using Console.  The second
-// also has a downside, in that any typing which occurs prior to invoking Console.ReadKey(true) will
-// be visible even though it wasn't supposed to be.  The downsides of the former approach are so large
-// and the cons of the latter minimal and constrained to the one API that we've chosen the second approach.
-// Thus, InitializeConsoleBeforeRead is called to set up the state of the console, then a read is done,
-// and then UninitializeConsoleAfterRead is called.
-void SystemNative_InitializeConsoleBeforeRead(uint8_t minChars, uint8_t decisecondsTimeout)
+void UninitializeConsole()
 {
-    struct termios newTermios;
-    if (tcgetattr(STDIN_FILENO, &newTermios) >= 0)
+    // This method is called on SIGQUIT/SIGINT from the signal dispatching thread
+    // and on atexit.
+
+    if (pthread_mutex_lock(&g_lock) == 0)
     {
-        if (!g_readInProgress)
+        if (g_consoleUninitialized)
+        {
+            return;
+        }
+
+        if (g_haveInitTermios)
         {
-            // Store the original settings, but only if we didn't already.  This function
-            // may be called when the process is resumed after being suspended, and if
-            // that happens during a read, we'll call this function to reset the attrs.
-            g_preReadTermios = newTermios;
+            // TODO: https://github.com/dotnet/cli/issues/6216
```

Comment:
@janvorli @stephentoub I think something is still needed here for https://github.com/dotnet/cli/issues/6216.
So I should be able to reproduce that issue with this implementation, but I'm not sure how.

Reply:
This has to do with tcsetattr stopping the process (default SIGTTOU) when the process is running in the background. I'm going to look a bit deeper into this.